gc/gc-base/src/main/java/org/projectnessie/gc/serialization/ReferencesKryoRegistrator.java

```diff
@@ -0,0 +1,92 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.spark.serializer.KryoRegistrator;
+import org.projectnessie.gc.base.ContentBloomFilter;
+import org.projectnessie.gc.base.ContentValues;
+import org.projectnessie.gc.base.IdentifiedResult;
+import org.projectnessie.gc.base.ImmutableGCParams;
+import org.projectnessie.model.ImmutableBranch;
+import org.projectnessie.model.ImmutableCommitMeta;
+import org.projectnessie.model.ImmutableDetached;
+import org.projectnessie.model.ImmutableIcebergTable;
+import org.projectnessie.model.ImmutableReferenceMetadata;
+import org.projectnessie.model.ImmutableTag;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ReferencesKryoRegistrator implements KryoRegistrator {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ReferencesKryoRegistrator.class);
+
+  /**
+   * Registering all needed classes. We need to register all used classes, because we are setting
+   * `spark.kryo.registrationRequired` to `true`. Setting it to false may impact the performance
+   * significantly, because as per doc:
+   *
+   * <p>Writing class names can cause significant performance overhead, so enabling this option can
+   * enforce strictly that a user has not omitted classes from registration.
+   *
+   * @see <a href="https://spark.apache.org/docs/latest/configuration.html">the Spark
+   *     documentation</a>
+   * @param kryo
+   */
+  @Override
+  public void registerClasses(Kryo kryo) {
+    LOGGER.info("Registering classes for kryo: " + kryo);
+    kryo.register(ImmutableBranch.class, new ImmutableBranchSerializer());
+    kryo.register(ImmutableDetached.class, new ImmutableDetachedSerializer());
+    kryo.register(ImmutableReferenceMetadata.class, new ImmutableReferenceMetadataSerializer());
+    kryo.register(ImmutableCommitMeta.class, new ImmutableCommitMetaSerializer());
+    kryo.register(ImmutableTag.class, new ImmutableTagSerializer());
+    kryo.register(ContentBloomFilter.class, new ContentBloomFilterSerializer());
+    kryo.register(ImmutableGCParams.class, new ImmutableGCParamsSerializer());
+    kryo.register(ImmutableIcebergTable.class);
+    kryo.register(IdentifiedResult.class);
+    kryo.register(ContentValues.class);
+
+    // needed for date-time
+    kryo.register(Duration.class);
+    kryo.register(Instant.class);
+
+    // needed for all types of collections. It will use serializers from
+    // `com.esotericsoftware.kryo.serializers.DefaultSerializers`
+    kryo.register(HashMap.class);
+    kryo.register(HashSet.class);
+    kryo.register(ConcurrentHashMap.class);
+    kryo.register(LinkedHashMap.class);
+    kryo.register(Collections.EMPTY_LIST.getClass());
+    kryo.register(Collections.EMPTY_MAP.getClass());
+    kryo.register(Collections.EMPTY_SET.getClass());
+    kryo.register(Collections.singletonList("").getClass());
```

Comment:
it is executed only once. Doing `Collections.singletonList("")` is a way to alleviate the problem that the inner class it private (same for two lines below)